src/transport/webrtc/substream.rs

```diff
@@ -18,132 +18,447 @@
+    Close,
+}
+
+/// Substream stream.
+enum State {
+    /// Substream is fully open.
+    Open,
 
-/// Channel-backed substream.
-#[derive(Debug)]
+    /// Remote is no longer interested in receiving anything.
+    SendClosed,
+}
+
+/// Channel-backedn substream.
 pub struct Substream {
-    /// Channel ID.
-    id: ChannelId,
+    /// Substream state.
+    state: Arc<Mutex<State>>,
 
-    /// TX channel for sending messages to transport.
-    tx: PollSender<(ChannelId, Vec<u8>)>,
+    /// Read buffer.
+    read_buffer: BytesMut,
 
-    /// RX channel for receiving messages from transport.
-    rx: ReceiverStream<Vec<u8>>,
+    /// TX channel for sending messages to `peer`.
+    tx: Sender<Event>,
 
-    /// Protocol codec.
-    codec: Option<ProtocolCodec>,
+    /// RX channel for receiving messages from `peer`.
+    rx: Receiver<Event>,
 }
 
 impl Substream {
     /// Create new [`Substream`].
-    pub fn new(id: ChannelId, tx: Sender<(ChannelId, Vec<u8>)>) -> (Self, Sender<Vec<u8>>) {
-        let (to_protocol, rx) = channel(64);
+    pub fn new() -> (Self, SubstreamHandle) {
+        let (outbound_tx, outbound_rx) = channel(256);
+        let (inbound_tx, inbound_rx) = channel(256);
+        let state = Arc::new(Mutex::new(State::Open));
+        let handle = SubstreamHandle {
+            tx: inbound_tx,
+            rx: outbound_rx,
+            state: Arc::clone(&state),
+        };
 
         (
             Self {
-                id,
-                codec: None,
-                tx: PollSender::new(tx),
-                rx: ReceiverStream::new(rx),
+                state,
+                tx: outbound_tx,
+                rx: inbound_rx,
+                read_buffer: BytesMut::new(),
             },
-            to_protocol,
+            handle,
         )
     }
+}
+
+/// Substream handle that is given to the transport backend.
+pub struct SubstreamHandle {
+    state: Arc<Mutex<State>>,
+
+    /// TX channel for sending messages to `peer`.
+    tx: Sender<Event>,
 
-    /// Apply codec for the substream.
-    pub fn apply_codec(&mut self, codec: ProtocolCodec) {
-        self.codec = Some(codec);
+    /// RX channel for receiving messages from `peer`.
+    rx: Receiver<Event>,
+}
+
+impl SubstreamHandle {
+    /// Handle message received from a remote peer.
+    ///
+    /// If the message contains any flags, handle them first and appropriately close the correct
+    /// side of the substream. If the message contained any payload, send it to the protocol for
+    /// further processing.
+    pub async fn on_message(&self, message: WebRtcMessage) -> crate::Result<()> {
+        if let Some(flags) = message.flags {
+            if flags == Flag::Fin as i32 {
+                let _ = self.tx.send(Event::RecvClosed).await?;
+            }
+
+            if flags & 1 == Flag::StopSending as i32 {
+                *self.state.lock() = State::SendClosed;
+            }
+
+            if flags & 2 == Flag::ResetStream as i32 {
```

Comment:
dq: We don't care about `Flag::FinAck` because we are closing the connection either way? (https://github.com/libp2p/specs/blob/master/webrtc/README.md#shared-concepts)

Reply:
I wonder why we apply binary `&`, even though flags seem to be exclusive and not a bitmask. Also, this way `FinAck` is handled by conditions `flags & 1 == StopSending` and `flags & 2 == ResetStream`, so this is probably the reason we don't have a separate condition for `FinAck`...

Reply:
I would merge as is and refactored this in a follow-up PR.